### PR TITLE
2.0.0

### PR DESCRIPTION
**Pull request by tohalla**
- Improve drift detection (when state changed outside terraform)
- Changes in storage devices (or partially in template) will not trigger recreation of the server resource)
- Only the template will be recreated when server resource is recreated
- The template is managed within server, when deleting a server, also the template storage will be destroyed. (previously used to delete all the other storages as well)
- The "normal" storage devices will be managed as separate resources (need to remove the declaration to actually remove the storage. Removing storage from within the server block just detaches it)
- The acceptance tests need additional work after the release
- Additional examples
- Improve documentation
- User agent header for terraform provider
- many of the utility functions are moved to /internal

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,37 +1,68 @@
...
- - Removed the plan, price, price_zone and timezone UpCloud resources
- - resource_upcloud_ip removed and replaced by resource_upcloud_floating_ip_address
- - resource_upcloud_firewall_rule removed and replaced by resource_upcloud_firewall_rules
- - resource_upcloud_zone removed and replaced by zone and zones datasources
+
+- Removed storage HCL blocks that failed due to referencing older UpCloud template ID
+- Removed the plan, price, price_zone and timezone UpCloud resources
+- resource_upcloud_ip removed and replaced by resource_upcloud_floating_ip_address
+- resource_upcloud_firewall_rule removed and replaced by resource_upcloud_firewall_rules
+- resource_upcloud_zone removed and replaced by zone and zones datasources
```
Comment (tohalla) [inline CHANGELOG.md:68]: Just formatting applied here. No change in older entries

File: .gitignore [hunk 2]
```diff
@@ -2,9 +2,6 @@
 terraform.tfstate*
 terraform-provider-upcloud
 crash.log
-*.tf
```
Comment (tohalla) [inline .gitignore:5]: examples should be included (no other tf files produced at any point)

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,37 +1,68 @@
...
- - Travis CI file execute to execute website-test covering provider documentation
- - Provider uses Terraform Plugin SDK V2
- - resource_upcloud_server expanded with new functionality from UpCloud API 1.3
- - resource_upcloud_storage expaned with new functionalty from UpCloud API 1.3
- - resource_upclopud_tag expanded to implement read function
- 
- 
+
+- README and examples/README to cover local builds, setup and test execution
+- Go verison to 1.14 and against Go master branch in Travis CI
```
Comment (ajmyyra) [inline CHANGELOG.md:55]: Just a typo here, verison -> version

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,37 +1,68 @@
...
+- Travis CI file execute to execute website-test covering provider documentation
+- Provider uses Terraform Plugin SDK V2
+- resource_upcloud_server expanded with new functionality from UpCloud API 1.3
+- resource_upcloud_storage expaned with new functionalty from UpCloud API 1.3
```
Comment (ajmyyra) [inline CHANGELOG.md:59]: Moar typos, functionalty -> functionality.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,37 +1,68 @@
...
- - Provider uses Terraform Plugin SDK V2
- - resource_upcloud_server expanded with new functionality from UpCloud API 1.3
- - resource_upcloud_storage expaned with new functionalty from UpCloud API 1.3
- - resource_upclopud_tag expanded to implement read function
- 
- 
+
+- README and examples/README to cover local builds, setup and test execution
+- Go verison to 1.14 and against Go master branch in Travis CI
+- Travis CI file execute to execute website-test covering provider documentation
```
Comment (ajmyyra) [inline CHANGELOG.md:56]: "execute to execute", hm?

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,37 +1,68 @@
...
+- Provider uses Terraform Plugin SDK V2
+- resource_upcloud_server expanded with new functionality from UpCloud API 1.3
+- resource_upcloud_storage expaned with new functionalty from UpCloud API 1.3
+- resource_upclopud_tag expanded to implement read function
```
Comment (ajmyyra) [inline CHANGELOG.md:60]: Typo here with `upcloud`.

File: upcloud/resource_upcloud_storage.go [hunk 7]
```diff
@@ -516,6 +464,41 @@ func resourceUpCloudStorageDelete(ctx context.Context, d *schema.ResourceData, m
...
+	storageDetails, err := client.GetStorageDetails(&request.GetStorageDetailsRequest{
+		UUID: d.Id(),
+	})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if len(storageDetails.ServerUUIDs) > 0 {
+		serverUUID := storageDetails.ServerUUIDs[0]
+		// Get server details for retrieven the address used to detach the storage
```
Comment (ajmyyra) [inline upcloud/resource_upcloud_storage.go:477]: Typo, retrieven -> retrieved.

**Review by ajmyyra — COMMENTED** [review]
Pointed out some typos but nothing major anywhere, good job! Only thing I wondered was the lack of support for more than 10 concurrent server creations (with Terraform waiting for first ones to complete before creating new ones). Did we have plans to support it in version 2 and/or was it scrapped?

**Comment by tohalla** [comment]
> Pointed out some typos but nothing major anywhere, good job! Only thing I wondered was the lack of support for more than 10 concurrent server creations (with Terraform waiting for first ones to complete before creating new ones). Did we have plans to support it in version 2 and/or was it scrapped?

Not relevant anymore. Dropped the mention about this

File: CHANGELOG.md (hunk 5 above)
Comment (tohalla) [reply CHANGELOG.md:56]: do not like poking around old changelog entries, but guess  I ll fix the typos 😁

**Review by ajmyyra — APPROVED** [review]

**Pull request merged**